### PR TITLE
EOS-16671 : add node functionality

### PR DESCRIPTION
**Pull request by shindeam**
## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    EOS-16671 : add node functionality
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
   YES
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Add node functionality to scale cluster
  </code>
</pre>
## Solution
<pre>
  <code>
    - Auth node before adding it to the cluster - ("pcs cluster auth <node> -u <username> -p <password>")
    - Checking the node count in cluster i.e. cluster size
    - If node count < 32 then adding node in normal way into the cluster ("pcs cluster node add <node>")
    - If node count > 32, failing the function & sending instructions back to use the remote-node add functionality
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    - print(cm.cluster_controller.add_node(nodeid="node3", cluster_user="hacluster", cluster_password="admin123"))
    - {"status": "InProgress", "msg": "Node node3 add operation started successfully in the cluster"}

    - print(cm.cluster_controller.add_node(nodeid="node2", cluster_user="hacluster", cluster_password="admin123"))
    - {"status": "Failed", "msg": "ClusterManagerException. add_node failed. Error: error(1): [run_cmd] Failed to execute pcs cluster node add node2 Error: Failed to execute pcs cluster node add node2 Error: Unable to add 'node2' to cluster: node is already in a cluster\n"}
  </code>
</pre>

### CODE REVIEW

File: ha/core/controllers/pcs/cluster_controller.py [hunk 1]
```diff
@@ -119,19 +134,41 @@ def storageset_list(self) -> dict:
...
-
+        :param cluster_user:
+        :param cluster_password:
+        :param node_id:
+        :param desc_file:
         Args:
-            nodeid (str, optional): Provide nodeid. Defaults to None.
-            filename (str, optional): Provide descfile. Defaults to None.
+            node_id (str, optional): Provide node_id. Defaults to None.
+            filename (str, optional): Provide desc_file. Defaults to None.
```
Comment (ajaysrivas) [inline ha/core/controllers/pcs/cluster_controller.py:147]: The description file handling should be done at upper layer. The cluster management functions should be implemented using standard parameters.

File: ha/core/controllers/pcs/cluster_controller.py [hunk 2]
```diff
@@ -119,19 +134,41 @@ def storageset_list(self) -> dict:
...
+
+        if desc_file:
+            _json_data = PcsClusterController.load_json_file(desc_file)
+            node_id = _json_data.get("node_id")
+            cluster_user = _json_data.get("cluster_user")
+            cluster_password = _json_data.get("cluster_password")
+        elif node_id and not(cluster_user or cluster_password):
+            return {"status": "Failed", "msg": "Missing parameters (cluster_user or cluster_password) for node_id"}
+
+        self.auth_node(node_id, cluster_user, cluster_password)
```
Comment (ajaysrivas) [inline ha/core/controllers/pcs/cluster_controller.py:164]: No error checking here. What if auth_node fails?

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

**Comment by ajaysrivas** [comment]
@shindeam Please fix the codacy issues.

File: ha/core/controllers/pcs/cluster_controller.py (hunk 2 above)
Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py]: We are checking error in auth_node itself and throwing ClusterManagerError which at the end is caught by our generic annotation @controller_error_handler

File: ha/core/controllers/pcs/cluster_controller.py (hunk 1 above)
Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py:220]: Updated

**Comment by shindeam** [comment]
> @shindeam Please resolve the conflicts and fix the Codacy issues.

Done

File: ha/core/controllers/pcs/pcs_controller.py [hunk 3]
```diff
@@ -72,6 +71,27 @@ def clean_failure_count(self, node_id):
...
+        Auth node to add
+        """
+        try:
+            _output, _err, _rc = self._execute.run_cmd(const.PCS_CLUSTER_NODE_AUTH.replace("<node>", node_id)
+                                                       .replace("<username>", cluster_user)
+                                                       .replace("<password>", cluster_password))
+        except Exception as e:
+            raise ClusterManagerError(f"Failed to authenticate node : {node_id} with reason : {e}")
+
+    def get_cluster_size(self):
```
Comment (ajay-paratmandali) [inline ha/core/controllers/pcs/pcs_controller.py:85]: convert it to protected function

from `get_cluster_size` to `_get_cluster_size`

File: ha/core/controllers/pcs/pcs_controller.py [hunk 4]
```diff
@@ -72,6 +71,27 @@ def clean_failure_count(self, node_id):
         _output, _err, _rc = self._execute.run_cmd(const.PCS_NODE_CLEANUP.replace("<node>", node_id),
                                                    check_error=False)
 
+    def auth_node(self, node_id, cluster_user, cluster_password):
```
Comment (ajay-paratmandali) [inline ha/core/controllers/pcs/pcs_controller.py:74]: convert it to protected function

from `auth_node` to `_auth_node`

File: ha/const.py [hunk 5]
```diff
@@ -97,6 +97,9 @@
 PCS_CLUSTER_START_NODE="pcs cluster start"
 PCS_CLUSTER_STATUS="pcs cluster status"
 PCS_CLUSTER_UNSTANDBY="pcs cluster unstandby --all"
+PCS_CLUSTER_NODE_AUTH="pcs cluster auth <node> -u <username> -p <password>"
+PCS_CLUSTER_NODE_ADD="pcs cluster node add <node> --start --enable"
+PCS_CLUSTER_PCSD_STATUS="sudo pcs status pcsd"
```
Comment (ajay-paratmandali) [inline ha/const.py:102]: i think sudo not needed

File: ha/core/controllers/pcs/cluster_controller.py [hunk 6]
```diff
@@ -63,6 +64,18 @@ def _pcs_cluster_start(self):
             Log.info(f"Pcs cluster start retry index : {retry_index}")
         return _status
 
+    @staticmethod
+    def load_json_file(json_file):
```
Comment (ajay-paratmandali) [inline ha/core/controllers/pcs/cluster_controller.py:68]: Move this to parent class

File: ha/core/controllers/pcs/pcs_controller.py (hunk 3 above)
Comment (shindeam) [reply ha/core/controllers/pcs/pcs_controller.py]: Done

File: ha/core/controllers/pcs/pcs_controller.py (hunk 4 above)
Comment (shindeam) [reply ha/core/controllers/pcs/pcs_controller.py]: Done

File: ha/const.py (hunk 5 above)
Comment (shindeam) [reply ha/const.py]: Done

File: ha/core/controllers/pcs/cluster_controller.py (hunk 6 above)
Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py]: Done

File: ha/core/controllers/pcs/cluster_controller.py (hunk 1 above)
Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py]: Done as per discussion removed descfile parameter and made nodeid, cluster_username & cluster_password as mandatory one

File: ha/core/controllers/pcs/cluster_controller.py (hunk 2 above)
Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py]: we are getting msg like below in response:

**Review by ajaysrivas — APPROVED** [review]

File: ha/core/controllers/pcs/cluster_controller.py (hunk 6 above)
Comment (ajay-paratmandali) [reply ha/core/controllers/pcs/cluster_controller.py]: move load json to pcs_controller not cluster controller
and rename load_json_file to _load_json_file

Comment (shindeam) [reply ha/core/controllers/pcs/cluster_controller.py]: Done

**Comment by gowthamchinna** [comment]
start vm-deployment

**Review by ajay-paratmandali — APPROVED** [review]

File: ha/core/controllers/pcs/pcs_controller.py [hunk 7]
```diff
@@ -72,6 +85,28 @@ def clean_failure_count(self, node_id):
...
+            _output, _err, _rc = self._execute.run_cmd(const.PCS_CLUSTER_NODE_AUTH.replace("<node>", node_id)
+                                                       .replace("<username>", cluster_user)
+                                                       .replace("<password>", cluster_password))
+        except Exception as e:
+            Log.error(f"Failed to authenticate node : {node_id} with reason : {e}")
+            raise ClusterManagerError(f"Failed to authenticate node : {node_id}, Please check username or password")
+
+    def _get_cluster_size(self):
+        """
+        Auth node to add
```
Comment (indrajitzagade) [inline ha/core/controllers/pcs/pcs_controller.py:102]: Please fix this.

**Pull request merged**